nltk/translate/ribes_score.py

```diff
@@ -0,0 +1,311 @@
+            worder.append(reference.index(h_word))
+        else:
+            # Note: range(1, max(i, hyp_len-i+1)) is the range of window sizes.
+            for window in range(1, max(i, hyp_len-i+1)):
+                if i+window < hyp_len: # If searching the right context is possible.
+                    # Retrieve the right context window.
+                    right_context_ngram = tuple(islice(hypothesis, i, i+window+1))
+                    num_times_in_ref = ref_ngrams.count(right_context_ngram)
+                    num_times_in_hyp = hyp_ngrams.count(right_context_ngram) 
+                    # If ngram appears only once in both ref and hyp.
+                    if num_times_in_ref == num_times_in_hyp == 1:
+                        # Find the position of ngram that matched the reference.
+                        pos = position_of_ngram(right_context_ngram, reference)
+                        worder.append(pos)  # Add the positions of the ngram.
+                        break
+                if window <= i: # If searching the left context is possible.
+                    # Retrieve the left context window.
+                    left_context_ngram = tuple(islice(hypothesis, i-window, i+1))
+                    num_times_in_ref = ref_ngrams.count(left_context_ngram)
+                    num_times_in_hyp = hyp_ngrams.count(left_context_ngram)
+                    if num_times_in_ref == num_times_in_hyp == 1:
+                        # Find the position of ngram that matched the reference.
+                        pos = position_of_ngram(left_context_ngram, reference)
+                        # Add the positions of the ngram.
+                        worder.append(pos+ len(left_context_ngram) -1)  
+                        break
+    return worder
+
+ 
+def choose(n, k):
+    """
+    This function is a fast way to calculate binomial coefficients, commonly
+    known as nCk, i.e. the number of combinations of n things taken k at a time. 
+    (https://en.wikipedia.org/wiki/Binomial_coefficient).
+    
+        >>> choose(4, 2)
+        6
+        >>> choose(6, 2)
+        15
+    
+    :param n: The number of things.
+    :type n: int
+    :param r: The number of times a thing is taken.
+    :type r: int
+    """
+    if 0 <= k <= n:
+        ntok, ktok = 1, 1
+        for t in range(1, min(k, n - k) + 1):
+            ntok *= n
+            ktok *= t
+            n -= 1
+        return ntok // ktok
+    else:
+        return 0
+
+    
+def find_increasing_sequences(worder):
+    """
+    Given the *worder* list, this function groups monotonic +1 sequences. 
+    
+        >>> worder = [7, 8, 9, 10, 6, 0, 1, 2, 3, 4, 5]
+        >>> list(find_increasing_sequences(worder))
+        [(7, 8, 9, 10), (0, 1, 2, 3, 4, 5)]
+    
+    :param worder: The worder list output from word_rank_alignment
+    :param type: list(int)
+    """
+    items = iter(worder)
+    a, b = None, next(items, None)
+    result = [b]
```

Comment:
If `worder == []`, find_increasing_sequence will not return None but an empty generator because it doesn't return a list but a generator. And the generator will be iterated later to calculate the kendall's tau, so when doing a `sum(())` it returns a 0. E.g.

```
>>> worder = []
>>> len(worder)
0
>>> def find_increasing_sequences(worder):
...     """
...     Given the *worder* list, this function groups monotonic +1 sequences. 
...     
...         >>> worder = [7, 8, 9, 10, 6, 0, 1, 2, 3, 4, 5]
...         >>> list(find_increasing_sequences(worder))
...         [(7, 8, 9, 10), (0, 1, 2, 3, 4, 5)]
...     
...     :param worder: The worder list output from word_rank_alignment
...     :param type: list(int)
...     """
...     items = iter(worder)
...     a, b = None, next(items, None)
...     result = [b]
...     while b is not None:
...         a, b = b, next(items, None)
...         if b is not None and a + 1 == b:
...             result.append(b)
...         else:
...             if len(result) > 1:
...                 yield tuple(result)
...             result = [b]
... 
>>> find_increasing_sequences([])
<generator object find_increasing_sequences at 0x7f9fbe71ca00>
>>> increasing_sequences = find_increasing_sequences([])
>>> sum(i for i in increasing_sequences)
0
>>> sum(())
```